### DISCUSSION TITLE
Hash Database lookup speeding up with bloom filter

### DISCUSSION THREAD

**Discussion by gfd2020** [sepinf-inc/IPED#2026] [Ideas]
If I'm not wrong, all analyzed files will have to be queried in the hashes database (when activated).
The problem is that most of them probably won't be found in the database. This way, a lot of time will be spent querying the database.
In this context, I imagined that we could use an optimization using bloom filter.

A Bloom filter is a space-efficient probabilistic data structure, conceived by Burton Howard Bloom in 1970, that is used to test whether an element is a member of a set. False positive matches are possible, but false negatives are not – in other words, a query returns either "possibly in set" or "definitely not in set".
This data structure is perfect for this case. We can avoid many hash database queries if we build this structure.

So I did a little test.

I took an IPHONE UFDR case with 20,000 files, 8 GB file.
I took NIST RDS IOS mininal hash db and transformed it to iped using iped-hashdb.jar.
The db hash sqlite file was 5GB ( 23363391entries ).
I used the guava bloom filter implementation, a lib that is already part of iped.
I put the md5 and sha1 hash in the bloom filter, generating a 37 MB file (the lib itself already has a serializer), very small compared to the original 5 GB.
I changed the HashDBDataSource class to read the 37 MB bloom filter file and in the 'lookup' method to test whether the hash to be tested would pass the bloom filter.
It was very simple, just do a test for each hash, if they all come back negative, you don't even need to consult the sqlite of the main iped db hash.
And the best thing is that guaja already has an implementation to store the information in bytes, so I didn't even need to transform the hash into a string.

Results:
In my 8GB test case, 600 files out of 20,000 will be identified by the NIST hash table.
Results obtained by the iped log file after processing is complete.

In the normal IPED implementation, 55 seconds were spent on the HashDBLookupTask task.
In the implementation with bloom filter, 0 seconds were spent on the HashDBLookupTask task.

Maybe there are some details that I haven't paid attention to, but I think this optimization would be very productive, if everything goes right.

**Comment by lfcnassif** [comment]
This seems a good optimization to me! But I'll defer the decision about integrating this to @wladimirleite, since he is the original author of HashDBLookupTask and knows all the details about it.

**Reply by gfd2020** [reply]
Yes, I completely agree.

**Reply by wladimirleite** [reply]
We use bloom filters in other projects, they are really useful.
I am out of the office until next Wednesday. I will take a closer look and reply here when I return.

**Comment by gfd2020** [comment]
Just a quick note.

1) The guava serializer performs good compression but is very slow to write and read the object. That's why I tested it with another serializer, kryo. The results were much better.

RDS (Modern PC + Android + IOS) minimal = 28 GB sqlite hash database, FPP 1%.

Guava serializer
Write object: 211 seconds
Read  object:  92 seconds
File size: 259 MB bloom filter

Kryo Serializer ( Deflate compress )
Write object: 12 seconds
Read  object:  3 seconds
File size: 267 MB bloom filter

Kryo Serializer ( without compress )
Write object: 3 seconds
Read  object: 0 seconds
File size: 291 MB bloom filter


2)
The false positive probability (FPP) of a bloom filter is defined as the probability that mightContain(Object). 
This parameter is configurable. The smaller the value, the fewer false positives but the size of the object increases. Default value is 3%.

Test Case: 5GB sqlite hash database. 20.000 files analyzed.
FPP = 5% -> 2339 possibly in set (1681 false positives ) - 34 MB bloom file
FPP = 3% -> 1514 possibly in set ( 856 false positives ) - 40 MB bloom file
FPP = 1% ->  866 possibly in set ( 208 false positives ) - 53 MB bloom file

**Reply by wladimirleite** [reply]
@gfd2020, did you measure how long does it take to generate the bloom filter (for the 28 GB SQLite)?

**Reply by gfd2020** [reply]
Yes, inside the database we have md5 and sha1 ( 113.373.675 hashs for each ), this took  178 seconds to read sqlite database and create the bloom filter object.
The writing and reading speeds to serialize this object are in the previous comment.

**Comment by gfd2020** [comment]
Small code for testing.

```
import com.google.common.hash.BloomFilter;
import com.google.common.hash.Funnels;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.sql.DriverManager;  
import java.sql.Connection;  
import java.sql.ResultSet;  
import java.sql.Statement;  
import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import org.objenesis.strategy.StdInstantiatorStrategy;
import com.esotericsoftware.kryo.util.DefaultInstantiatorStrategy;
import com.esotericsoftware.kryo.util.Pool;
import java.io.*;
import java.util.zip.*;

public class GuavaBloomFilter {
    public static void main(String[] args) throws Exception {
	String sqlitePath = "teste.sqlite";
	double fpp = 0.01;
	BloomFilter<byte[]> bloomFilter = generateBF(sqlitePath,fpp);
        timeBF(bloomFilter);
    }

    private static BloomFilter<byte[]> generateBF(String sqlitePath, double fpp) {

		BloomFilter<byte[]> bf = null;
		try {

			String url = "jdbc:sqlite:"+sqlitePath;
			Connection conn = DriverManager.getConnection(url);  		
			Statement stmt  = conn.createStatement();  			
			String sqlCount = "SELECT count(*) as count FROM HASHES";  
			ResultSet rs    = stmt.executeQuery(sqlCount);  
			long records = 0;
			while (rs.next()) {  
				records = rs.getLong("count");
			}
			String sql = "SELECT MD5,SHA1 FROM HASHES";  
			rs    = stmt.executeQuery(sql);  
			records *= 2;
			bf = BloomFilter.create(Funnels.byteArrayFunnel(),records,fpp);
			
			long startTime = System.currentTimeMillis();		  
			long k = 0;
			long per = 0;
			long lastper = -1;
			while (rs.next()) {  
				bf.put(rs.getBytes("MD5"));				
				bf.put(rs.getBytes("SHA1"));
				k ++;
				per = (long)(k*100/records*2);
				if (per != lastper){
					System.out.print("\rReading sqlite and creation bloom filter: "+(per)+"%");
					lastper = per;
				}
			}  
			System.out.print("\r                                                  ");
			System.out.print("\r");
			long endTime = System.currentTimeMillis();
			System.out.println("Read sqlite and creation time : " + (endTime - startTime)/1000 + " seconds");			
		
		}catch (Exception e){
			System.out.println(e.getMessage()); 
		}		
        return bf;
    }

    public static void timeBF(BloomFilter<byte[]> bf) throws IOException {
		
	long startTime = System.currentTimeMillis();
        FileOutputStream fos = new FileOutputStream("guava.bloom");
        bf.writeTo(fos);
        fos.close();
	long endTime = System.currentTimeMillis();			
        System.out.println("Write guava object : " + (endTime - startTime)/1000 + " seconds");
		
        startTime = System.currentTimeMillis();
        FileInputStream fis = new FileInputStream("guava.bloom");
        BloomFilter<byte[]> tb = BloomFilter.readFrom(fis, Funnels.byteArrayFunnel());
        endTime = System.currentTimeMillis();
	tb = null;
        System.out.println("Read guava object  : " + (endTime - startTime)/1000 + " seconds");
		
	startTime = System.currentTimeMillis();
	kryoSerialize("kryo.bloom",bf,false);
	endTime = System.currentTimeMillis();
	System.out.println("Write kryo object uncompressed : " + (endTime - startTime)/1000 + " seconds");
		
	startTime = System.currentTimeMillis();
	tb = kryoDeserialize("kryo.bloom",BloomFilter.class,false);		
        endTime = System.currentTimeMillis();
	tb = null;
        System.out.println("Read kryo object uncompressed  : " + (endTime - startTime)/1000 + " seconds");	

	startTime = System.currentTimeMillis();
	kryoSerialize("kryo.bloom_zip",bf,true);
	endTime = System.currentTimeMillis();
	System.out.println("Write kryo object compressed   : " + (endTime - startTime)/1000 + " seconds");
		
	startTime = System.currentTimeMillis();
	tb = kryoDeserialize("kryo.bloom_zip",BloomFilter.class,true);
        endTime = System.currentTimeMillis();
	tb = null;
        System.out.println("Read kryo object compressed    : " + (endTime - startTime)/1000 + " seconds");

    }

	private static void kryoSerialize(String file, Object obj, boolean compress) {
		Kryo kryo = kryoPool.obtain();
		try {
			Output output = null;
			if (compress){
				FileOutputStream fos = new FileOutputStream(file);
				DeflaterOutputStream dos = new DeflaterOutputStream(fos);		
				output = new Output(dos);
			}else{
				FileOutputStream fos = new FileOutputStream(file);
				output = new Output(fos);		
			}		
			kryo.writeObject(output, obj);
			output.close();
		} catch (Exception e) {
			System.out.println(e.getMessage()); 
		} finally {
			kryoPool.free(kryo);
		}
	}

	private static <T> T kryoDeserialize(String file, Class<T> clazz, boolean compress) {
		Kryo kryo = kryoPool.obtain();
		try {
			kryo.setInstantiatorStrategy(new DefaultInstantiatorStrategy(new StdInstantiatorStrategy()));			
			Input input = null;
			if(compress){
				FileInputStream fis = new FileInputStream(file);
				InputStream in = new InflaterInputStream(fis);		
				input = new Input(in);
			}else{
				File initialFile = new File(file);
				InputStream targetStream = new FileInputStream(initialFile);
				input = new Input(targetStream);
			}			
			Object ret = kryo.readObject(input, clazz);
			input.close();
			return (T) ret;
		} catch (Exception e) {
			System.out.println(e.getMessage()); 
			return null;
		} finally {
			kryoPool.free(kryo);
		}
	}

	static Pool<Kryo> kryoPool = new Pool<Kryo>(true, false, 8) {
		@Override
		protected Kryo create() {
			Kryo kryo = new Kryo();
			kryo.setReferences(true);
			kryo.setRegistrationRequired(false);
			return kryo;
		}
	};   	
}

```

**Comment by wladimirleite** [comment]
Hello @gfd2020!

You mentioned that a case with ~20K files HashDBLookupTask took 55 seconds (with a ~5GB database), right? 
Did you mean that the processed case has only 20K items?
I am using a much larger database (~45 GB), and I checked a few processed cases I have here and time spent HashDBLookupTask seems lower.
A case with 2.4M items took 95 seconds in HashDBLookupTask.
Another with 3.5M items took 129 seconds.

Another observation is about the "hit ratio". Obviously it depends a lot on the evidence files being processed and the hashes database content. Checking ~20 processed cases in the local server we use here, on average 37% of the items were found in the database.

I will try to run the small program you provide to check how long it will take to process the database I have here and how much memory the filters will require.

**Reply by gfd2020** [reply]
@wladimirleite 

>You mentioned that a case with ~20K files HashDBLookupTask took 55 seconds (with a ~5GB database), right?
Did you mean that the processed case has only 20K items?

Yes.

> A case with 2.4M items took 95 seconds in HashDBLookupTask.

Very fast. 

I did this 55 second test on the master 4.2 snapshot version. I ran the same test in versions 4.1.2 and 4.1.4 (on another machine, too) and HashDBLookupTask was very fast, taking 5 second or less to execute.
I couldn't figure out why there was a difference. I didn't imagine there could be such a big difference for the stable version.

I would like to apologize for taking up your time with this. I think there's a lesson for me. I'll look into the reason for this difference. I apologize once again and I also ask that you don't waste any more of your time on this. I'll check what happened.

**Reply by wladimirleite** [reply]
No need to apologize, at all!  
I was using master version for the cases that I mentioned, but the master from a couple of weeks ago.
In theory, the time taken by HashDBLookupTask should not change between master and 4.1.x, as there was no code change related to it (as far as I remember).

**Reply by gfd2020** [reply]
Ok, I got it. Anyway, I'll see what happened.

**Reply by gfd2020** [reply]
Very strange. I did several tests. In stable and master versions. It's intermittent. Sometimes the HashDBLookupTask performs very poorly, other times the performance is normal. Will there be a specific race condition? I will continue to do tests.

**Reply by wladimirleite** [reply]
As it uses the SQLite JDBC driver, which uses native code and shared buffers, there is some unpredictable concurrence with other threads accessing other SQLite files (e.g. parsers). 
That may explain small variations, but I never noticed a major slow down.

Is the hashes database in the same disk used as temp or case output? 
It is on a SSD? (In the past I ran tests on HDDs, and it is much slower and the performance variation is much higher).

**Reply by gfd2020** [reply]
Iped binaries, hash database, temp and case output in the same SSD. 
Evidence location on same SSD and another tests on  another disk ( External HDD USB 3.0 ), same odd results.

![image](https://github.com/sepinf-inc/IPED/assets/59742865/9a2b7867-b875-40fe-9316-5fa48d73697e)

**Reply by wladimirleite** [reply]
Ideally, temp should be on another disk, but that shouldn't case such huge variation.

**Reply by gfd2020** [reply]
If it's a driver problem, there's not much we can do. I'll test with temp on another disk. In any case, the bloom filter can help reduce requests to the database, I don't know if it will pay off.

**Reply by lfcnassif** [reply]
Maybe SO disk caches on RAM are involved. I often use RamMap app from SysInternals to clear SO caches when I want to benchmark changes in IO bound tasks in a evidence first processing scenario, running it multiple times.

**Reply by gfd2020** [reply]
Thanks for the rammap tip. But clearing the cache will harm the performance of HashDBLookupTask, right? I did some tests and it seems that after clearing the cache, HashDBLookupTask's performance is very bad. It seems like that's the issue...

**Reply by wladimirleite** [reply]
It shouldn't (at least not by that much). Also, after the first queries cache should be warmed up.
I will try to reproduce this using RamMap tomorrow.

**Reply by gfd2020** [reply]
Ok. I used RDS IOS Mininal for hash db and a UFDR IPHONE evidence. Just found 650 pictures on RDS hash.

**Comment by lfcnassif** [comment]
> But clearing the cache will harm the performance of HashDBLookupTask, right?

Not sure, but I think so, if performance is always bad after clearing the cache, it possibly is the cause of the fluctuations you are experiencing. For large cases, like @wladimirleite's it shouldn't make a huge difference since the hash DB will be warmed up like he said. For small cases, like yours, clearing the cache can make a bigger difference. A last tip, the option inside RamMap to clear SO caches is "Clear Stand By Lists"

PS1: Using a separate disk for the hash DB may be better to benchmark it, avoiding interferences from other things.

PS2: Clearing caches when SSDs are used should make smaller differences.

**Comment by wladimirleite** [comment]
I ran more tests with a smaller UFDR (55K items), and cleared system caches before some of the processing:
Clear -> Process 1 -> Process  2 -> Process 3 -> Clear -> Process 4
It doesn't make any noticeable difference, in all 4 tests HashDBLookupTask took 1 second.

**Reply by lfcnassif** [reply]
Sorry guys for pointing to a misleading direction. If hash DB is on SSD, SO cache shouldn't make a big difference like @wladimirleite anticipated and tested.

**Comment by wladimirleite** [comment]
@gfd2020, I made some minor modifications to the sample program you provided, as we support 5 types of hashes (4 are currently used in our latest database version) and the number of records is not the same for each type:
```text
MD5      150,573,274
SHA1     138,262,652
SHA256     3,748,854
EDONKEY    3,748,854
SHA512             0
```

Here is the output:
```text
Count hashes time : 10 seconds
Total records : 296333634
Read sqlite and creation time : 543 seconds
Write guava object : 206 seconds
Read guava object  : 88 seconds
Write kryo object uncompressed : 1 seconds
Read kryo object uncompressed  : 0 seconds
Write kryo object compressed   : 18 seconds
Read kryo object compressed    : 4 seconds
```

So it would take ~9 minutes to update the bloom filter when a new version of the hashes database is used.
Using the kyro uncompressed serializer, loading the filter would be very fast.
The filter object would consume ~350 MB.

My conclusions are: the 9 minutes delay seems a bit too long, but acceptable to me, as it would not be frequent. About the memory, 350 MB is not that much, but could be inconvenient for some setups with less RAM. So I don't see any red flag, but I am not sure if it pays off.

```java
    private static BloomFilter<byte[]> generateBF(String sqlitePath, double fpp) {

        BloomFilter<byte[]> bf = null;
        try {
            long startTime0 = System.currentTimeMillis();

            String url = "jdbc:sqlite:" + sqlitePath;
            Connection conn = DriverManager.getConnection(url);
            Statement stmt = conn.createStatement();
            String[] hashes = {"MD5","SHA1","SHA256","SHA512","EDONKEY"};
            long records = 0;
            for (String h : hashes) {
                String sqlCount = "SELECT count(*) as count FROM HASHES WHERE " + h + " IS NOT NULL";
                ResultSet rs = stmt.executeQuery(sqlCount);
                while (rs.next()) {
                    records += rs.getLong("count");
                }
                rs.close();
            }
            long endTime0 = System.currentTimeMillis();
            System.out.println("Count hashes time : " + (endTime0 - startTime0) / 1000 + " seconds");
            System.out.println("Total records : " + records);

            long startTime1 = System.currentTimeMillis();
            String sql = "SELECT ";
            for (int i = 0; i < hashes.length; i++) {
                if (i > 0) sql += ",";
                sql += hashes[i];
            }
            sql += " FROM HASHES";
            ResultSet rs = stmt.executeQuery(sql);
            bf = BloomFilter.create(Funnels.byteArrayFunnel(), records, fpp);
            long k = 0;
            long per = 0;
            long lastper = -1;
            while (rs.next()) {
                for (String h : hashes) {
                    byte[] b = rs.getBytes(h);
                    if (b != null) {
                        bf.put(b);
                        k++;
                    }
                }
                per = (long) (k * 100 / records);
                if (per != lastper) {
                    System.out.print("\rReading sqlite and creation bloom filter: " + (per) + "%");
                    lastper = per;
                }
            }
            System.out.print("\r                                                  ");
            System.out.print("\r");
            long endTime1 = System.currentTimeMillis();
            System.out.println("Read sqlite and creation time : " + (endTime1 - startTime1) / 1000 + " seconds");

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        return bf;
    }
```

**Reply by gfd2020** [reply]
Hi @wladimirleite . Thank you for the tests. I think it's best to leave this update on hold.

Just make another test with a bigger evidence. 32 GB UFDR report - 140.000 files.

Cleaned cache.
[INFO]  [engine.core.Statistics]                        HashDBLookupTask:       Processing Time:        734s (31%)

Running again ( cache warmed )
[INFO]  [engine.core.Statistics]                        HashDBLookupTask:       Processing Time:        600s (29%)

Another test with HDD disk image, 250 GB , windows machine.
[engine.core.Statistics]                        HashDBLookupTask:       Processing Time:        1228s (26%)

---------------

I think that if all the recommendations for using the hash database are followed, the performance is very good. However, I believe that I will not be able to meet all of them in my environment, in which case I will have to implement the bloom filter to increase the performance of my cases. Another option would be not to use RDS, since the objective would be to reduce the time but it is getting longer.

I would like to thank you for your help in clarifying.

**Reply by wladimirleite** [reply]
It would be better to understand what is going on before taking any decision.
Please, when you have some time, try placing IPED temporary folder on another SSD (different from the hashes database).

**Reply by lfcnassif** [reply]
@gfd2020 is hash DB folder (and temp) excluded from anti-virus scanning?

**Reply by gfd2020** [reply]
Nop ..

**Comment by patrickdalla** [comment]
Isn't bloom filter by definition based on hash? I. e. hash lookup is alredy
a kind of bloom filter. And MD5 or sha256 removes any probabilistic feature
(entropy) of data, so, it seems that bloom filter to search on md5 or
sha256 set won't improve overall performance.

Em qua., 13 de dez. de 2023 11:27, gfd2020 ***@***.***>
escreveu:

**Reply by wladimirleite** [reply]
The point is that the hashes (MD5, SH1 etc.) use a lot of space, so they are stored in a SQLite. 
The Bloom Filter of such hash sets would be much smaller, so it could be stored in memory, so the lookup would be faster.

**Reply by gfd2020** [reply]
Exactly. 
And in practice I achieved a lot of performance improvement when hashlookudbtask did not have good performance.
